src/lib.rs

```diff
@@ -382,22 +384,20 @@ macro_rules! cmd {
 /// ```
 #[derive(Debug, Clone)]
 pub struct Shell {
-    cwd: RefCell<PathBuf>,
-    env: RefCell<HashMap<OsString, OsString>>,
+    cwd: Arc<Path>,
+    env: HashMap<Arc<OsStr>, Arc<OsStr>>,
```

Comment:
Hmm that's interesting, Then in a `Cmd` if you modify either the `cwd` or the `env` you'll CoW it via `Arc::get_mut`.

As for keeping the parent, are you imagining some API that exposes it somehow? or do you want the env to be applied recursively? (that'll probably cost more than cloning the full env)

Reply:
> or do you want the env to be applied recursively? (that'll probably cost more than cloning the full env)

This is what I had in mind, but, yeah, thats probably overkill, and its better to just clone the hash map itself. 

The good news is that ultimately it doesn't matter, because in Rust muttable and persistant data structures have the same API.